### ISSUE TITLE
Required and Default

### ISSUE THREAD

**Issue by nardil** [OAI/OpenAPI-Specification#1543]
Hi all, i need a clarification about using Required and Default properties.

Consider this definition took from OpenAPI specs "Models with Polymorphism Support" example:
```
type: object
properties:
  huntingSkill:
    type: string
    description: The measured skill for hunting
    default: lazy
    enum:
      - clueless
      - lazy
      - adventurous
      - aggressive
required:
  - huntingSkill
```
If `huntingSkill` property required, i'm expected to provide a value or get a validation error, so `default` property is useless. Otherwise, if is supposed to use default value if not provided, it's `required` property to be useless.

Can you help clarify?

Thank you!

**Comment by handrews** [comment]
@nardil `default` documents *behavior* when a field is absent.  It does not cause the default value to be written in to an instance, although of course an application can do that and many tools support doing so automatically.  But it has no impact on validation.

`required` is a validation keyword that is evaluated against the instance as it exists when validation is invoked.  Again, some validators will, for convenience, provide the option to write the default in just before validation.  This is just because walking a schema and writing defaults into the instance involves much the same code as validation, so it's easier to implement in a validator.  But it is ***NOT*** part of validation.

The main reason to use `required` and `default` is if your application will pre-populate a user input form with the default value, and you want to prevent the user from removing the value entirely.  "form" here is not necessarily a web form- it could be that a code generator uses a programming language defaulting feature in a method signature, for example.

Most of the time, `default` is used with optional fields to document that they behave _as if they were set to_ a certain value when he field is missing.  Note also that `default` applies to *missing* fields.  A field set to `null` is not missing, as `null` is a value in JSON.

**Comment by hkosova** [comment]
@nardil 
It looks like `default` was supposed to be removed from that example in https://github.com/OAI/OpenAPI-Specification/issues/1077 and PR [1089](https://github.com/OAI/OpenAPI-Specification/pull/1089/files). According to @darrelmiller:

> I am recommending that we remove the `default` because it risks making the `required` notion invalid and encourages clients to unnecessarily send `default` values over the wire.

... but that change did not make it into the 3.0.0 Specification (or was reverted?).

<br/>
@handrews

> The main reason to use `required` and `default` is if your application will pre-populate a user input form with the default value, and you want to prevent the user from removing the value entirely.

Unless I'm misunderstanding, this seems like a use case for the OpenAPI [`example`](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.1.md#user-content-schemaExample) keyword - to provide an _example value_ for documentation or display purposes.

> it could be that a code generator uses a programming language defaulting feature in a method signature, for example.

If a parameter is declared as having a default value, this usually (always?) implies that the parameter is _optional_ rather than required. In other words, using `default` with `required` somewhat contradicts the meaning of `required`, as noted in @darrelmiller's comment above.

**Comment by handrews** [comment]
@hkosova `default` and `required` together is definitely odd.  I was pointing out that potential use case because it's come up in the JSON Schema spec repo.  I doubt it's common.

Modern JSON Schema has both `examples` (note the plural) and `default`.  They have distinct purposes, although implementations MAY treat the `default` as an additional example (or as the only example if `examples` is not present).

`examples` is "here are some things you can do, some of which may be unusual which is why they are being called out"
`default` is generally "here is the thing that happens if you don't do anything"

All annotation keywords (including `default` and `examples`) simply provide information to applications without mandating whether or how it must be used.  This is why the initial UI form value use case is valid- it's basically an indirect form of writing it into the instance.

**Comment by handrews** [comment]
Last answer was thumbs-upped, so closing.

**Comment by oralokan** [comment]
@handrews You mentioned that using `default` and `required` together is odd. Apparently, [Swagger Documentation](https://swagger.io/docs/specification/describing-parameters/) asserts that this one of the two common mistakes when using the `default` keyword.

However, I'm having some difficulty reconciling that statement with the scenario that I'll describe below. I would appreciate help in understanding why this isn't an issue, or if it is, how it can be resolved.

Suppose we are implementing a web server. As with all servers, it has to know which port to listen to. Hence, `port` is a `required` property of the server's _runtime configuration_.

On the other hand, 80 is the default port for web servers. It stands to reason that 80 can be used as the default value unless user explicitly specifies a different port in the _user configuration_, i.e. `port` is an `optional` property of the server's _user configuration_.

One might say we are dealing with two distinct schemas: the user configuration (the configuration explicitly submitted by the user) and the runtime configuration (the actual configuration used by the server). Then,

- `port` is an optional parameter of the user configuration with a default value of 80.
- `port` is a required parameter of the runtime configuration.
- Resolving the user configuration by applying default values yields the runtime configuration.

My issue is, how do we justify having to define two separate schemas while setting both `default` and `required` on the `port` property of a single `ServerConfiguration` schema seems to achieve the same result? Indeed it does on some validator implementations, but I've come across some that don't support this citing that the spec doesn't include it.

I don't think this scenario is that far-fetched or odd... perhaps I am missing something? Any clarification would be much appreciated.

**Comment by handrews** [comment]
Aw... `default`, no... 😜 

More seriously, we need to take a bit of a step back to explain this.

There's an architectural problem with `default`.  JSON Schema does not include modifying instance data while evaluating the schema+instance.  The intention is that (as is the case in 2020-12 and therefore OAS 3.1), annotations can be collected during evaluation and included in standard machine-readable output.  (BTW, this is why we say "evaluating" instead of just "validating", because evaluating includes both validating and annotating.)

But... annotations can only be attached to instance locations that exist, and `default` is primarily meaningful if the instance location _doesn't_ exist.  So, the official JSON Schema answer to "why can't we just write things into the instance" is "you should just collect the annotations and use them to change the instance however you need to", but that actually doesn't work for `default`, which is the main keyword people want to use to change the instance. 🤦 

There's a proposal to deprecate `default` and add a keyword called `missing` or `propertyDefaults` or something like that (naming is hard) that would work like `required`:  it would sit at the object level, so that it shows up as an annotation as long as the object is present at all.  Then you could actually evaluate a schema+instance, get the annotation output, and write the missing/propertyDefaults values in before further processing.

You would not be able to mark that property `required`, but you could structure your application to ensure that defaults are all filled in.  Or if you only want some defaults filled in, you could add a custom annotation for that (the next release of JSON Schema will clarify that you should be able to add custom annotations without having to write code to support the keyword, since JSON Schema already knows how to collect annotations).

Of course, none of this does you any good in 3.0 or earlier, but there's not much we can do about that.

**Comment by oralokan** [comment]
@handrews, thanks a lot for your detailed answer.

> Aw... default, no... 😜

I guess it just goes to show there is real demand for it out there 😀. The proposal to add a keyword at the object level sounds promising.

> You would not be able to mark that property required

Is it really necessary to prevent people from using both? Let me briefly explain why I want to be able to use them together.

Continuing on my earlier example, suppose we are developing a web server using TypeScript. Suppose we define `Configuration` in an OpenAPI 3.0 document as:

``` yaml
Configuration:
    type: object
    required:
        - port
        - bindAddress
    properties:
        port:
            type: number
            default: 80
        bindAddress:
            type: string
            default: "0.0.0.0"
```

Suppose we use something like [openapi-typescript](https://github.com/drwpow/openapi-typescript) to generate TypeScript type annotations corresponding to this schema:

``` typescript
type Configuration = {
    port: number;
    bindAddress: string;
}
```

Notice that neither property of the generated type is optional, which is because we specified `required` for both of them. After all, both of these properties are in fact required for a runtime configuration to be valid.

Using [Ajv](https://ajv.js.org) with the `useDefaults` option in this context works like a charm:

``` typescript
// schema is the JSON Schema for Configuration
function isValidConfiguration(config:unknown, schema:any): config is Configuration {
    const ajv = new Ajv({useDefaults:true});
    const valid = ajv.validate(schema, config);
    if (!valid) {
        console.log("Instance does not conform to schema", ajv.errors);
        return false;
    }
    return true;
}

const config = loadConfiguration() || {};
const schema = loadSchema();

if (isValidConfiguration(config, schema)) {
    // at this point, missing values have been filled in by Ajv
    startServer(config);
} else {
    exit();
}
```

> but you could structure your application to ensure that defaults are all filled in. 

Being able to delegate this to a validator like Ajv _greatly_ simplifies the application code. Basically,

- I understand that `required` and `default` are not intended to be used together, but doing so is not prohibited (at least in 3.0).
- Doing just that in combination with a validator that modifies the instance and a type generator that generates static types leads to what I found to be a highly streamlined and efficient approach to validation.
- While I can do this with Ajv, I cannot do it with other validators. I suppose that's to be expected seeing that this matter has been debated considerably and there don't seem to be any perfect solutions.

I suspect it might be difficult to continue using the same approach if a future version of the OpenAPI spec explicitly prohibits the new `missing` or `propertyDefaults` or whatever it ends up being used together with `required`.

What do you think? Might there be some justification for using them together after all?

**Comment by handrews** [comment]
I would not support anything forbidding the use of keywords together.  JSON Schema doesn't do that.  Some keywords don't make sense together, but we don't forbid using them that way (in part because there may be reasons that are not yet obvious to use the combination).

AJV is problematic in several ways.  The most severe being its default "strict mode" which is egregiously non-compliant with the spec.  The writing-in-default-values is not inherently problematic, but introduces a new processing dynamic with significant implications for what it means to be a compliant JSON Schema implementation.

The annotation approach (have a `"missing"` keyword and use it to do the modification post-schema evaluation) fits within the current processing model and solves most problems.

No one has really made a coherent case in the JSON Schema project to support instance modification as part of evaluation.  It hasn't necessarily been rejected, but it's not clear that the set of use cases that would solve compared to the annotation-based approach is worth the added complexity.

**Comment by oralokan** [comment]
Thank you very much @handrews for taking the time to respond so quickly with thorough answers.

> I would not support anything forbidding the use of keywords together. JSON Schema doesn't do that. Some keywords don't make sense together, but we don't forbid using them that way (in part because there may be reasons that are not yet obvious to use the combination).

That's reassuring and good to know.

> No one has really made a coherent case in the JSON Schema project to support instance modification as part of evaluation. It hasn't necessarily been rejected, but it's not clear that the set of use cases that would solve compared to the annotation-based approach is worth the added complexity.

Do you think support for type generation for statically typed languages (as in the example with TypeScript) could be a possible case? As I understand, it might be because of Ajv's non-compliance that I'm able to do what I'm doing.

**Comment by handrews** [comment]
As long as you turn strict mode off (it's on by default, at least in the versions of AJV that support 2019-09 and 2020-12, not sure about earlier) and don't rely on extensions, AJV is fine.  Although sadly it does not support interoperable extensions using vocabularies, only its own non-portable extension mechanism.

We've never tried to forbid implementations from writing in default values, but as you noted it's not portable (and if it's enabled by default, that's definitely non-compliant).

Code generation (and generative use cases as whole, whether we're talking about code, docs, UIs, etc.) is an interesting case because it's not using the evaluation model of a schema + an instance.  So it doesn't get into the difficulty where the evaluation model can't see `"default"` for a missing instance location.  You're just walking the schema and can do whatever you want with it.

I'm firmly of the belief that we (the JSON Schema Org) need to address generative use cases more directly, but it has not happened yet.

**Comment by oralokan** [comment]
@handrews 

> Code generation (and generative use cases as whole, whether we're talking about code, docs, UIs, etc.) is an interesting case because it's not using the evaluation model of a schema + an instance. So it doesn't get into the difficulty where the evaluation model can't see "default" for a missing instance location. You're just walking the schema and can do whatever you want with it.
> 
> I'm firmly of the belief that we (the JSON Schema Org) need to address generative use cases more directly, but it has not happened yet.

I certainly think that would be a great idea. In my personal experience, I have found that being able to use type definitions made in the OpenAPI document to be very useful. By using schemas defined there for all validation tasks, the OpenAPI document becomes a part of the code rather than separate documentation that has to be constantly revised. Also using these schemas to generate types reduces duplicity and ensures that validated types are used across the codebase.

I look forward to any developments in OpenAPI that address generative use cases more directly.

**Comment by oralokan** [comment]
A quick wrap-up.

I have come to agree with the OpenAPI documentation that states that `required` and `default` do not make sense together. `required` states that the data submitted by the user _must_ include the property, otherwise it is invalid. So for valid data, the `default` value will never be used. If the `default` value was to be applied prior to validation, well in that case the user was not really _required_ to include the property in the first place.

The main reason I wanted to use `required` and `default` together was due to code generation (in my case, generating TypeScript types). Setting `required` ensures that the property is a required property of the generated type.

However, being required in use submitted data and being required in the generated type are two separate things.

Specifically, if the JSON schema includes a `default` value, then deserialized objects will _always_ include a value for that property: either the user-submitted value or the default value.

As such, I think the correct solution is to ensure that the generated type includes the property as a required property either if it is marked as `required` in the JSON Schema, or it has a default value.

Basically, this has to be implemented in the code-generation library.

Having said that, it would be great if OpenAPI specifies how code-generation should work so that we can have compliant code-generation libraries.
